manager/controlapi/ca_rotation_test.go

```diff
@@ -0,0 +1,771 @@
+				SigningCACert: rotationCert,
+				ExternalCAs: []*api.ExternalCA{
+					{
+						URL:    initExtServer.URL,
+						CACert: initialLocalRootCA.CACert, // current cert
+					},
+					{
+						URL:    rotateExtServer.URL,
+						CACert: rotationCert, //new cert
+					},
+				},
+			},
+			expectErrorString: "rotating from one external CA to a different external CA is not supported",
+		},
+		{
+			rootCA: initialLocalRootCA,
+			caConfig: api.CAConfig{
+				SigningCACert: append(rotationCert, initialLocalRootCA.CACert...),
+				SigningCAKey:  rotationKey,
+			},
+			expectErrorString: "cannot contain multiple certificates",
+		},
+		{
+			rootCA: initialLocalRootCA,
+			caConfig: api.CAConfig{
+				SigningCACert: testutils.ReDateCert(t, rotationCert, rotationCert, rotationKey,
+					time.Now().Add(-1*time.Minute), time.Now().Add(2*helpers.OneYear)),
+				SigningCAKey: rotationKey,
+			},
+			expectErrorString: "expires too soon",
+		},
+		{
+			rootCA: initialLocalRootCA,
+			caConfig: api.CAConfig{
+				SigningCACert: initialLocalRootCA.CACert,
+				SigningCAKey:  testutils.ExpiredKey, // same cert but mismatching key
+			},
+			expectErrorString: "certificate key mismatch",
+		},
+		{
+			// this is just one class of failures caught by NewRootCA, not going to bother testing others, since they are
+			// extensively tested in NewRootCA
+			rootCA: initialLocalRootCA,
+			caConfig: api.CAConfig{
+				SigningCACert: testutils.ExpiredCert,
+				SigningCAKey:  testutils.ExpiredKey,
+			},
+			expectErrorString: "expired",
+		},
+	} {
+		cluster := &api.Cluster{
+			RootCA: invalid.rootCA,
+			Spec: api.ClusterSpec{
+				CAConfig: invalid.caConfig,
+			},
+		}
+		secConfig := getSecurityConfig(t, &localRootCA, cluster)
+		_, err := validateCAConfig(context.Background(), secConfig, cluster)
+		require.Error(t, err, invalid.expectErrorString)
+		require.Equal(t, codes.InvalidArgument, grpc.Code(err), invalid.expectErrorString)
+		require.Contains(t, grpc.ErrorDesc(err), invalid.expectErrorString)
+	}
+}
+
+func runValidTestCases(t *testing.T, testcases []*rootCARotationTestCase, localRootCA *ca.RootCA) {
+	for _, valid := range testcases {
+		cluster := &api.Cluster{
+			RootCA: *valid.rootCA.Copy(),
+			Spec: api.ClusterSpec{
+				CAConfig: valid.caConfig,
+			},
+		}
+		secConfig := getSecurityConfig(t, localRootCA, cluster)
+		result, err := validateCAConfig(context.Background(), secConfig, cluster)
+		require.NoError(t, err, valid.description)
+
+		// ensure that the cluster was not mutated
+		require.True(t, reflect.DeepEqual(valid.rootCA, cluster.RootCA))
+
+		fmtString := "%s:\ncluster value:\n%v\nexpected value:\n%v\n"
+
+		// Because join tokens are random, we can't predict exactly what it is, so this needs to be manually checked
+		require.Equal(t, valid.expectJoinTokenChange, !reflect.DeepEqual(result.JoinTokens, valid.rootCA.JoinTokens),
+			fmtString, valid.description, result.JoinTokens, valid.rootCA.JoinTokens)
```

Comment:
Slightly more verbose, but I think this would be easier to read and more useful as:

```go
if valid.expectJoinTokenChange {
        require.NotEqual(t, result.JoinTokens, valid.rootCA.JoinTokens, fmtString, valid.description, result.JoinTokens, valid.rootCA.JoinTokens)
} else {
        require.Equal(t, result.JoinTokens, valid.rootCA.JoinTokens, fmtString, valid.description, result.JoinTokens, valid.rootCA.JoinTokens)
}
```